File: 0886-score-of-parentheses/0886-score-of-parentheses.java
class Solution {
    public int scoreOfParentheses(String s) {
        Deque<Integer> stack = new ArrayDeque<>();
        stack.push(0);

        for (char c : s.toCharArray()) {
            if (c == '(') stack.push(0);
            else {
                int cur = stack.pop();
                int score = (cur == 0) ? 1 : cur * 2;
                int prev = stack.pop();
                stack.push(prev + score);
            }
        }

        return stack.pop();
    }
}

File: 0886-score-of-parentheses/0886-score-of-parentheses.kt
class Solution {
    fun scoreOfParentheses(s: String): Int {
        val stack = ArrayDeque<Int>()
        stack.addLast(0)

        for (c in s) {
            when (c) {
                '(' -> stack.addLast(0)
                ')' -> {
                    val cur = stack.removeLast()
                    val score  = if (cur == 0) 1 else cur * 2
                    stack.addLast(stack.removeLast() + score)
                }
            }
        }

        return stack.last()
    }
}